src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -192,6 +192,358 @@ public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount,
+            if (index < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charIndex, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            if (count < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charCount, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            if (chars.Length - index < count)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.chars, ExceptionResource.ArgumentOutOfRange_IndexCountBuffer);
+            Contract.EndContractBlock();
+
+            byte[] bytes;
+            if (count > 0) {
+                fixed (char* input = chars)
+                    bytes = GetBytesAsciiFastPath(encoding, input + index, count);
+            } else {
+                bytes = Array.Empty<byte>();
+            }
+
+            return bytes;
+
+        }
+
+        public unsafe static int GetBytesAsciiFastPath(Encoding encoding, char[] chars, int charIndex, int charCount, byte[] bytes, int byteIndex)
+        {
+            // Fast path for pure ASCII data for ASCII and UTF8 encoding
+            Debug.Assert(encoding != null);
+            if (chars == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.chars, ExceptionResource.ArgumentNull_Array);
+            if (bytes == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.bytes, ExceptionResource.ArgumentNull_Array);
+            if (charIndex < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charIndex, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            if (charCount < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charCount, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            if (chars.Length - charIndex < charCount)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.chars, ExceptionResource.ArgumentOutOfRange_IndexCountBuffer);
+            if (byteIndex < 0 || byteIndex > bytes.Length)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.byteIndex, ExceptionResource.ArgumentOutOfRange_Index);
+            Contract.EndContractBlock();
+
+            // Note that byteCount is the # of bytes to decode, not the size of the array
+            int byteCount = bytes.Length - byteIndex;
+            if (charCount > 0 && byteCount == 0)
+                ThrowBytesOverflow(encoding);
+
+            int lengthEncoded;
+            if (charCount > 0 && byteCount > 0) {
+                fixed (char* pInput = chars)
+                fixed (byte* pOutput = &bytes[0]) {
+                    char* input = pInput + charIndex;
+                    byte* output = pOutput + byteIndex;
+                    lengthEncoded = GetBytesAsciiFastPath(input, output, Math.Min(charCount, byteCount));
+                    if (lengthEncoded < byteCount) {
```

Comment:
should we check if lengthEncoded is >= charCount and then exit? your check here can cause calling encoding.GetBytesFallback while is not needed (like in cases having pure ascii string and calling with big byte array parameter